### ISSUE TITLE
Compilation error about type inferrence

### ISSUE THREAD

**Issue by vorner** [acmcarther/cucumber-rs#50]
Hello

If I try to compile it with recent nightly (rustc 1.24.0-nightly (5a2465e2b 2017-12-06)) or newest stable, it fails with bunch of these messages:

```
error: code relies on type inference rules which are likely to change
   --> features/step_definitions/project_steps.rs:111:3
    |
111 | /   Then!(c,
112 | |         "^the feature reports no undefined steps$",
113 | |         |_, world: &mut CucumberWorld, _| {
114 | |     match world.execute_result {
...   |
120 | |     }
121 | |   });
    | |_____^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #39216 <https://github.com/rust-lang/rust/issues/39216>
    = note: this error originates in a macro outside of the current crate
```

Considering the kind of error and the fact the last commit is about a year ago, is this project still alive and maintained? Or, is there a replacement? Is it worth trying to send pull requests?

**Comment by acmcarther** [comment]
Interesting. Just from looking at it, I'd guess that specifying types for arg (1) and arg (3) would alleviate the problem, but it does negatively impact the ergonomics. 

I don't use this library on any recent projects that I've been working on though I'd love to get back to it. I'm up for fixing this, or reviewing pull requests that do. There are probably some good low hanging fruit as well to be fixed at the same time:
- Update cargo deps
- Update bundler/ruby cucumber.

**Comment by vorner** [comment]
I'll have a look if I could tweak the API somewhat and come up with something that works without the specification. But I'm afraid I won't get to it that soon, there are too many other crates I want to improve too and a day job in addition to that 😇.

Actually, it would be cool if it could be done in this way:

```rust
#[cucumber("^the feature reports no undefined steps")]
fn no_undef_steps(_:(), world: &mut CucumberWold, _: ()) { ... }
```

But I'm afraid this would require nightly for now. Still, there could be better ways to do it ergonomically. I might try experimenting with it.

**Comment by acmcarther** [comment]
Yeah there is a lot of potential for proc-macros to help here by generating glue code. I haven't written any proc-macros myself yet... its definitely something on my list.
